src/emit/pretty_printer/doc.cpp

```diff
@@ -35,41 +40,79 @@ auto Doc::operator+(const Doc &other) const -> Doc
 
 auto Doc::operator&(const Doc &other) const -> Doc
 {
-    return *this + space() + other;
+    return *this + Doc::text(" ") + other;
 }
 
 auto Doc::operator/(const Doc &other) const -> Doc
 {
     return *this + line() + other;
 }
 
-auto Doc::nest(int indent) const -> Doc
+auto Doc::operator|(const Doc &other) const -> Doc
 {
-    return Doc(makeNest(indent, impl_));
+    return *this + hardline() + other;
 }
 
-auto Doc::group() const -> Doc
+auto Doc::operator<<(const Doc &other) const -> Doc
 {
-    return Doc(makeUnion(flatten(impl_), impl_));
+    // *this + (line() + other).nest(DEFAULT_INDENT)
+    auto nested = Doc(makeNest(makeConcat(line().impl_, other.impl_)));
+    return *this + nested;
 }
 
-// Rendering
-auto Doc::render(int width) const -> std::string
+auto Doc::hardIndent(const Doc &other) const -> Doc
 {
-    Renderer renderer(width);
-    return renderer.render(impl_);
+    // *this + (hardline() + other).nest(DEFAULT_INDENT)
+    auto nested = Doc(makeNest(makeConcat(hardline().impl_, other.impl_)));
+    return *this + nested;
 }
 
-// Internal access
-auto Doc::impl() const -> const std::shared_ptr<DocImpl> &
+// Compound assignment operators
+auto Doc::operator+=(const Doc &other) -> Doc &
 {
-    return impl_;
+    *this = *this + other;
+    return *this;
 }
```

Comment:
Can't we just do
```suggestion
{
    return (*this += other);
}
```
1. parenthesis not needed but makes intent clear (though not really sure if it's actually cleaner than the initial approach)

Reply:
bruh how do you want to use += within its own definition

Reply:
lol